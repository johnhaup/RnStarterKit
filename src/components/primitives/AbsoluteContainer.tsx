import { get, isEmpty } from 'lodash';
import React from 'react';
import { StyleSheet, View, ViewStyle } from 'react-native';

interface Props {
  children?: JSX.Element | JSX.Element[];
  top?: boolean | number;
  bottom?: boolean | number;
  left?: boolean | number;
  right?: boolean | number;
  center?: boolean;
  style?: ViewStyle | ViewStyle[];
  pointerEvents?: 'auto' | 'box-none' | 'none' | 'box-only';
}

const positionProps = ['top', 'bottom', 'left', 'right'];

export const AbsoluteContainer = ({
  children,
  pointerEvents = 'box-none',
  center,
  style = {},
  ...rest
}: Props) => {
  const positionProperties = positionProps.reduce((a, c) => {
    const value = get(rest, c, undefined);
    if (value) {
      return {
        ...a,
        [c]: typeof value === 'number' ? value : 0,
      };
    }
    return a;
  }, {});
  const centerStyles = center ? styles.center : null;

  const containerStyle = isEmpty(positionProperties)
    ? [StyleSheet.absoluteFill, centerStyles, style]
    : [styles.base, positionProperties, centerStyles, style];
  return (
    <View style={containerStyle} pointerEvents={pointerEvents}>
      {children}
    </View>
  );
};

const styles = StyleSheet.create({
  base: { position: 'absolute' },
  t: {
    top: 0,
  },
  b: {
    bottom: 0,
  },
  l: {
    left: 0,
  },
  r: {
    right: 0,
  },
  center: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  empty: {},
});
