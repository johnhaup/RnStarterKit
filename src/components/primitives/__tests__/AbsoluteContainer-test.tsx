/**
 * @format
 */

import React from 'react';
import 'react-native';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { AbsoluteContainer } from '../AbsoluteContainer';

it('renders correctly', () => {
  const snap = renderer.create(<AbsoluteContainer />);
  expect(snap).toMatchSnapshot();
});
