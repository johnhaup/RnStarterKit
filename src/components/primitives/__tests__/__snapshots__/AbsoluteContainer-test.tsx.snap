// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Array [
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      null,
      Object {},
    ]
  }
/>
`;
